### PR TITLE
[SYCL] Support for read_write access mode for accessors with target

### PR DESCRIPTION
**Pull request by garimagu**
host_image.
- tests to be added after scheduler support is added for image class.

Signed-off-by: Garima Gupta <garima.gupta@intel.com>

### CODE REVIEW

File: sycl/include/CL/sycl/accessor.hpp [hunk 1]
```diff
@@ -275,9 +282,12 @@ class image_accessor
...
   static_assert(
-      AccessMode == access::mode::read || AccessMode == access::mode::write ||
-          AccessMode == access::mode::discard_write,
-      "Access modes can be only read/write/discard_write for image accessor.");
+      ((IsImageAcc || IsImageArrayAcc) &&
+       (IsImageAccessWriteOnly || IsImageAccessReadOnly)) ||
+          (IsHostImageAcc && (IsImageAccessAnyWrite || IsImageAccessAnyRead)),
+      "Access modes can be only read/write/discard_write for image/image_array "
+      "target accessor, or they can be only "
+      "read/write/discard_write/read_write for host_image target accessor.");
```
Comment (garimagu) [inline sycl/include/CL/sycl/accessor.hpp]: The assert is essentially checking:
If the accessor is Image Acc Or Image_array Acc, then Access mode can be WriteOnly or ReadOnly.
Or
If the accessor is HostImage Acc, then Access mode can be AnyWrite or AnyRead.
Where the AnyWrite,AnyRead, WriteOnly and ReadOnly are image_accessor defined.

**Comment by garimagu** [comment]
Changes here correspond to the update in SYCL spec, posted in PR:
https://github.com/KhronosGroup/SYCL-Docs/pull/15

File: sycl/include/CL/sycl/accessor.hpp [hunk 2]
```diff
@@ -382,24 +392,29 @@ class image_accessor
   };
 
   //  Available only when: (accessTarget == access::target::image ||
-  //  accessTarget == access::target::host_image) && accessMode ==
-  //  access::mode::read
+  //  accessTarget == access::target::host_image) && (accessMode ==
+  //  access::mode::read || accessMode == access::mode::read_write)
```
Comment (bader) [inline sycl/include/CL/sycl/accessor.hpp]: I think the right condition is:
```c++
(accessTarget == access::target::image && accessMode == access::mode::read) || 
(accessTarget == access::target::host_image && (accessMode == access::mode::read || accessMode == access::mode::read_write))
```

File: sycl/include/CL/sycl/accessor.hpp [hunk 3]
```diff
@@ -382,24 +392,29 @@ class image_accessor
...
+  //  The appropriate allowed AccessMode for the AccessTarget is checked in the
+  //  static_assert in the class.
   template <typename CoordT, int Dims = Dimensions,
             typename = detail::enable_if_t<
                 (Dims > 0) && (IsValidCoordDataT<Dims, CoordT>::value) &&
-                (IsImageAcc || IsHostImageAcc) && IsImageAccessRead>>
+                (IsImageAcc || IsHostImageAcc) && IsImageAccessAnyRead>>
```
Comment (bader) [inline sycl/include/CL/sycl/accessor.hpp]: This allows `read_write` access on the device, which is not supposed to be allowed.

File: sycl/include/CL/sycl/accessor.hpp [hunk 4]
```diff
@@ -409,6 +424,8 @@ class image_accessor
   // Available only when: (accessTarget == access::target::image || accessTarget
   // == access::target::host_image) && accessMode == access::mode::write ||
   // accessMode == access::mode::discard_write
```
Comment (bader) [inline sycl/include/CL/sycl/accessor.hpp]: Please, update the condition in the comment.

File: sycl/include/CL/sycl/accessor.hpp [hunk 5]
```diff
@@ -382,24 +392,29 @@ class image_accessor
...
+  //  Available only when: (accessTarget == access::target::image ||
+  //  accessTarget == access::target::host_image) && (accessMode ==
+  //  access::mode::read || accessMode == access::mode::read_write)
+  //  The appropriate allowed AccessMode for the AccessTarget is checked in the
+  //  static_assert in the class.
   template <typename CoordT, int Dims = Dimensions,
             typename = detail::enable_if_t<
                 (Dims > 0) && (IsValidCoordDataT<Dims, CoordT>::value) &&
-                (IsImageAcc || IsHostImageAcc) && IsImageAccessRead>>
+                (IsImageAcc || IsHostImageAcc) && IsImageAccessAnyRead>>
```
Comment (bader) [inline sycl/include/CL/sycl/accessor.hpp]: `read_write` access mode must be available with the host accessor.

File: sycl/include/CL/sycl/accessor.hpp (hunk 2 above)
Comment (garimagu) [reply sycl/include/CL/sycl/accessor.hpp]: That is true. The static assert at line 284 takes care that read_write is not used with image targets.
So here the condition checked is as you have written.
Let me know if you still would like me to change this.

Comment (bader) [reply sycl/include/CL/sycl/accessor.hpp]: Yes. I still think it should be fixed as this comment is misleading.

File: sycl/include/CL/sycl/accessor.hpp (hunk 5 above)
Comment (garimagu) [reply sycl/include/CL/sycl/accessor.hpp]: added the requested changes.

File: sycl/include/CL/sycl/accessor.hpp (hunk 4 above)
Comment (garimagu) [reply sycl/include/CL/sycl/accessor.hpp]: added the requested changes.

File: sycl/include/CL/sycl/accessor.hpp (hunk 3 above)
Comment (garimagu) [reply sycl/include/CL/sycl/accessor.hpp]: added the requested changes.

File: sycl/include/CL/sycl/accessor.hpp [hunk 6]
```diff
@@ -381,39 +391,49 @@ class image_accessor
         T, detail::type_list<cl_int4, cl_float4>>::type::value;
   };
 
-  //  Available only when: (accessTarget == access::target::image ||
-  //  accessTarget == access::target::host_image) && accessMode ==
-  //  access::mode::read
+  // Available only when:
+  // (accessTarget==access::target::image && accessMode==access::mode::read) ||
+  // (accessTarget==access::target::host_image &&
+  //  accessMode==access::mode::read_write)
```
Comment (bader) [inline sycl/include/CL/sycl/accessor.hpp]:
```suggestion
  //  (accessMode == access::mode::read || accessMode == access::mode::read_write))
```

File: sycl/include/CL/sycl/accessor.hpp [hunk 7]
```diff
@@ -381,39 +391,49 @@ class image_accessor
         T, detail::type_list<cl_int4, cl_float4>>::type::value;
   };
 
-  //  Available only when: (accessTarget == access::target::image ||
-  //  accessTarget == access::target::host_image) && accessMode ==
-  //  access::mode::read
+  // Available only when:
+  // (accessTarget==access::target::image && accessMode==access::mode::read) ||
```
Comment (bader) [inline sycl/include/CL/sycl/accessor.hpp]: Please, add space around binary operators.
```suggestion
  // (accessTarget == access::target::image && accessMode == access::mode::read) ||
```

File: sycl/include/CL/sycl/accessor.hpp [hunk 8]
```diff
@@ -381,39 +391,49 @@ class image_accessor
...
+  // (accessTarget==access::target::host_image &&
```
Comment (bader) [inline sycl/include/CL/sycl/accessor.hpp]:
```suggestion
  // (accessTarget == access::target::host_image &&
```

File: sycl/include/CL/sycl/accessor.hpp [hunk 9]
```diff
@@ -381,39 +391,49 @@ class image_accessor
...
   DataT read(const CoordT &Coords) const {
     // TODO: To be implemented.
     throw cl::sycl::feature_not_supported("Read API is not implemented.");
     return;
   };
 
-  // Available only when: (accessTarget == access::target::image || accessTarget
-  // == access::target::host_image) && accessMode == access::mode::read
+  // Available only when:
+  // (accessTarget==access::target::image && accessMode==access::mode::read) ||
```
Comment (bader) [inline sycl/include/CL/sycl/accessor.hpp]:
```suggestion
  // (accessTarget == access::target::image && accessMode == access::mode::read) ||
```

File: sycl/include/CL/sycl/accessor.hpp [hunk 10]
```diff
@@ -381,39 +391,49 @@ class image_accessor
...
+  // (accessTarget==access::target::host_image &&
```
Comment (bader) [inline sycl/include/CL/sycl/accessor.hpp]:
```suggestion
  // (accessTarget == access::target::host_image &&
```

File: sycl/include/CL/sycl/accessor.hpp [hunk 11]
```diff
@@ -381,39 +391,49 @@ class image_accessor
...
+  //  accessMode==access::mode::read_write)
```
Comment (bader) [inline sycl/include/CL/sycl/accessor.hpp]:
```suggestion
  //  (accessMode == access::mode::read || accessMode == access::mode::read_write))
```

File: sycl/include/CL/sycl/accessor.hpp [hunk 12]
```diff
@@ -381,39 +391,49 @@ class image_accessor
...
     // TODO: To be implemented.
     throw cl::sycl::feature_not_supported("Read API is not implemented.");
     return;
   };
 
-  // Available only when: (accessTarget == access::target::image || accessTarget
-  // == access::target::host_image) && accessMode == access::mode::write ||
-  // accessMode == access::mode::discard_write
+  // Available only when:
+  // (accessTarget==access::target::image &&
```
Comment (bader) [inline sycl/include/CL/sycl/accessor.hpp]:
```suggestion
  // (accessTarget == access::target::image &&
```

File: sycl/include/CL/sycl/accessor.hpp [hunk 13]
```diff
@@ -381,39 +391,49 @@ class image_accessor
...
+  // (accessMode==access::mode::write ||
```
Comment (bader) [inline sycl/include/CL/sycl/accessor.hpp]:
```suggestion
  // (accessMode == access::mode::write ||
```

File: sycl/include/CL/sycl/accessor.hpp [hunk 14]
```diff
@@ -381,39 +391,49 @@ class image_accessor
...
+  //  accessMode==access::mode::discard_write)) ||
```
Comment (bader) [inline sycl/include/CL/sycl/accessor.hpp]:
```suggestion
  //  accessMode == access::mode::discard_write)) ||
```

File: sycl/include/CL/sycl/accessor.hpp [hunk 15]
```diff
@@ -381,39 +391,49 @@ class image_accessor
...
+  // (accessTarget==access::target::host_image &&
```
Comment (bader) [inline sycl/include/CL/sycl/accessor.hpp]:
```suggestion
  // (accessTarget == access::target::host_image &&
```

File: sycl/include/CL/sycl/accessor.hpp [hunk 16]
```diff
@@ -381,39 +391,49 @@ class image_accessor
...
+  // (accessMode==access::mode::write || accessMode==access::mode::discard_write
```
Comment (bader) [inline sycl/include/CL/sycl/accessor.hpp]:
```suggestion
  // (accessMode == access::mode::write || accessMode == access::mode::discard_write
```

File: sycl/include/CL/sycl/accessor.hpp [hunk 17]
```diff
@@ -381,39 +391,49 @@ class image_accessor
...
+  //  || accessMode==access::mode::read_write))
```
Comment (bader) [inline sycl/include/CL/sycl/accessor.hpp]:
```suggestion
  //  || accessMode == access::mode::read_write))
```

File: sycl/include/CL/sycl/accessor.hpp (hunk 6 above)
Comment (garimagu) [reply sycl/include/CL/sycl/accessor.hpp]: done

File: sycl/include/CL/sycl/accessor.hpp (hunk 7 above)
Comment (garimagu) [reply sycl/include/CL/sycl/accessor.hpp]: done

File: sycl/include/CL/sycl/accessor.hpp (hunk 8 above)
Comment (garimagu) [reply sycl/include/CL/sycl/accessor.hpp]: done

File: sycl/include/CL/sycl/accessor.hpp (hunk 10 above)
Comment (garimagu) [reply sycl/include/CL/sycl/accessor.hpp]: done

File: sycl/include/CL/sycl/accessor.hpp (hunk 11 above)
Comment (garimagu) [reply sycl/include/CL/sycl/accessor.hpp]: done

File: sycl/include/CL/sycl/accessor.hpp (hunk 12 above)
Comment (garimagu) [reply sycl/include/CL/sycl/accessor.hpp]: done

File: sycl/include/CL/sycl/accessor.hpp (hunk 13 above)
Comment (garimagu) [reply sycl/include/CL/sycl/accessor.hpp]: done

File: sycl/include/CL/sycl/accessor.hpp (hunk 14 above)
Comment (garimagu) [reply sycl/include/CL/sycl/accessor.hpp]: done

File: sycl/include/CL/sycl/accessor.hpp (hunk 15 above)
Comment (garimagu) [reply sycl/include/CL/sycl/accessor.hpp]: done

File: sycl/include/CL/sycl/accessor.hpp (hunk 16 above)
Comment (garimagu) [reply sycl/include/CL/sycl/accessor.hpp]: done

File: sycl/include/CL/sycl/accessor.hpp (hunk 17 above)
Comment (garimagu) [reply sycl/include/CL/sycl/accessor.hpp]: done

File: sycl/include/CL/sycl/accessor.hpp (hunk 6 above)
Comment (bader) [reply sycl/include/CL/sycl/accessor.hpp]: NOTE: `accessMode == access::mode::read` check was not added as suggested.

File: sycl/include/CL/sycl/accessor.hpp (hunk 11 above)
Comment (bader) [reply sycl/include/CL/sycl/accessor.hpp]: NOTE: `accessMode == access::mode::read` check was not added as suggested.

File: sycl/include/CL/sycl/accessor.hpp (hunk 6 above)
Comment (garimagu) [reply sycl/include/CL/sycl/accessor.hpp]: added the comment and uploaded it again. 
Sorry about that. I changed the logic in the code, missed to add it in the comment.

**Pull request merged**